### DISCUSSION TITLE
Response Schema, Fastapi Streaming Responses  for Documentation

### DISCUSSION THREAD

**Discussion by anantgupta129** [fastapi/fastapi#11552] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, HTTPException
from fastapi.responses import StreamingResponse
from pydantic import BaseModel

import json
import time
import asyncio

app = FastAPI()

class ErrorModel(BaseModel):
    detail: str

def fake_json_streamer():
    t0 = time.time()
    for i in range(10):
        print(f"Chunk being yielded (time {int((time.time()-t0)*1000)}ms)", flush=True)
        yield json.dumps( {"message": "Hello World"}) + '\n'
        time.sleep(0.5)
    print(f"Over (time {int((time.time()-t0)*1000)}ms)", flush=True)


@app.get("/test", responses={500: {"model": ErrorModel}})
async def test():
    try:
        return StreamingResponse(fake_json_streamer(), media_type='text/event-stream')
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"An error occurred while processing the request. {str(e)}")
```

### Description

How to add schema for documentation to return streaming responses for text or json 

### Operating System

Windows

### FastAPI Version

0.109.2

### Pydantic Version

2.6.1

### Python Version

3.9.19

**Comment by n8sty** [comment]
See the additional responses section of the docs https://fastapi.tiangolo.com/advanced/additional-responses/

**Reply by anantgupta129** [reply]
but it does not cover any model for streaming response, how can i add response model for streaming responses so that its correctly displayed in the swagger/ redoc documentation

**Reply by YuriiMotov** [reply]
Could you give any examples or references to the openapi specification on what you want to achieve?

**Reply by anantgupta129** [reply]
hi i want to have clearer documentation, when accesing /docs or /redoc, what is the best to mention that the request is a streaming response

**Reply by YuriiMotov** [reply, marked as answer]
Hi!

If you are talking about clarifications for people who read docs, then you can add the description for `200` response:
```
@app.get("/test", responses={200: {"description": "Successfull response will return a stream of JSON objects"}, 500: {"model": ErrorModel}})
```

And if you are talking about specifying response type for client code generators, it requires more investigation..
Have you read discussions below?
https://github.com/OAI/OpenAPI-Specification/issues/1576
https://github.com/OAI/OpenAPI-Specification/issues/3730
https://github.com/OAI/OpenAPI-Specification/pull/3769

**Comment by osangu** [comment]
Hi there! I know this question has already been resolved and quite some time has passed, but I just wanted to leave a comment because I think I’ve found an answer that’s closest to what the original poster might have intended. I’ve confirmed that it works as follows with fastapi==0.115.13 and pydantic==2.11.7.

```python
from typing import Literal
from pydantic import BaseModel
from fastapi.responses import StreamingResponse

class StreamExampleResponse(BaseModel):
    event: Literal['your', 'event']
    data: 'ExampleElement'

    class ExampleElement(BaseModel):
        your: int
        field: str

    def model_dump_sse(self):
        return f'event: {self.event}\ndata: {self.data.model_dump_json()}\n\n'


@app.get(
    path='/sse',
    responses={
        200: {
            'content': {
                'text/event-stream': {
                    'schema': StreamExampleResponse.model_json_schema()
                }
            }
        }
    },
    response_class=StreamingResponse
)
async def ex_sse_view_function():
    def ex_context():
        for i in range(10):
            response = StreamExampleResponse(
                event='your',
                data={'your': i, 'field': str(i) + ' field'}
            )

            yield response.model_dump_sse()

    return StreamingResponse(
        ex_context(),
        media_type='text/event-stream'
    )
```

<img width="1442" height="796" alt="image" src="https://github.com/user-attachments/assets/897cba0d-3fb3-4b43-a082-63151135b50c" />
<img width="1390" height="280" alt="스크린샷 2025-07-25 오후 3 59 15" src="https://github.com/user-attachments/assets/d873b0b6-fb12-4829-ac2a-475eafc74764" />

**Reply by gaardhus** [reply]
Wouldn't it be nice if you could do like this, or am I missing something?

```python
from typing import Literal
from pydantic import BaseModel
from fastapi.responses import StreamingResponse

class StreamExampleResponse(BaseModel):
    event: Literal['your', 'event']
    data: 'ExampleElement'

    class ExampleElement(BaseModel):
        your: int
        field: str

    def model_dump_sse(self):
        return f'event: {self.event}\ndata: {self.data.model_dump_json()}\n\n'


@app.get(path='/sse')
async def ex_sse_view_function() -> StreamingResponse[StreamExampleResponse]:
    def ex_context():
        for i in range(10):
            response = StreamExampleResponse(
                event='your',
                data={'your': i, 'field': str(i) + ' field'}
            )

            yield response.model_dump_sse()

    return StreamingResponse(
        ex_context(),
        media_type='text/event-stream'
    )
```
